doesn't build on 6.9 kernel

Zenergy lo longer works on linux 6.9 mainline build, while it was working at 6.8. 

So I tried to investigate why and decided to compile the module manually, as I was previously installing it from AUR. 
But when trying to build this module in linux 6.9, it gives the following error:

```
/home/mparagames/zenergy/zenergy.c:229:27: error: ‘struct cpuinfo_x86’ has no member named ‘x86_max_cores’
  229 |         sockets = cpus / c->x86_max_cores;
      |
```

as it can be seen in this image:

![image](https://github.com/BoukeHaarsma23/zenergy/assets/68190118/e40fedfc-6ea5-4845-9845-2fe121083fcc)

however, apparently linux 6.9 changed the way information about the amount of physical cores, threads (logical cores), and dies per socket is obtained, according to this changelog on the documentation:

![image](https://github.com/BoukeHaarsma23/zenergy/assets/68190118/7549fd10-2ce6-4a10-b9c8-10d5b71e55cc)

so I tried changing line 229 of the code to 

`sockets = cpus / topology_num_cores_per_package();`

replacing c->x86_max_cores to topology_num_cores_per_package() will make it compile on linux 6.9, but i assume it would then fail on linux 6.8 and below (haven't tested it, however).

![image](https://github.com/BoukeHaarsma23/zenergy/assets/68190118/69fbf2a0-891c-4a1d-98c8-9caf2102c942)

However keep in mind that I am no kernel or kernel module developer, I am just a cs student and I only know very basic c and code some projects as a hobby (usually in high level languages). Therefore, please be aware that I am prone to mistakes, as I don't really know what exactly would be the side effects of this change. 
So i just kindly ask that you look deeper at what caused the change in linux 6.9 that made zenergy stop working, and I am giving this as a possible hint of what could be going on from what I observed. 

Thank you in advance.

--- a/zenergy.c
+++ b/zenergy.c
@@ -24,6 +24,7 @@
 #include <linux/slab.h>
 #include <linux/topology.h>
 #include <linux/types.h>
+#include <linux/version.h>
 
 #define DRV_MODULE_DESCRIPTION	"AMD energy driver"
 #define DRV_MODULE_VERSION	"1.0"
@@ -207,7 +208,9 @@ static int amd_create_sensor(struct device *dev,
 			     enum hwmon_sensor_types type, u32 config)
 {
 	struct hwmon_channel_info *info = &data->energy_info;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 9, 0)
 	struct cpuinfo_x86 *c = &boot_cpu_data;
+#endif
 	struct sensor_accumulator *accums;
 	int i, num_siblings, cpus, sockets;
 	u32 *s_config;
@@ -223,10 +226,15 @@ static int amd_create_sensor(struct device *dev,
 	cpus = num_present_cpus() / num_siblings;
 
 	/*
-	 * c->x86_max_cores is the linux count of physical cores
+	 * topology_num_cores_per_package (or c->x86_max_cores prior to 6.9) is
+	 * the linux count of physical cores.
 	 * total physical cores/ core per socket gives total number of sockets.
 	 */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 9, 0)
 	sockets = cpus / c->x86_max_cores;
+#else
+	sockets = cpus / topology_num_cores_per_package();
+#endif
 
 	s_config = devm_kcalloc(dev, cpus + sockets + 1,
 				sizeof(u32), GFP_KERNEL);

